nav2_map_server/src/occ_grid_loader.cpp

```diff
@@ -152,25 +160,66 @@ nav2_util::CallbackReturn OccGridLoader::on_configure(const rclcpp_lifecycle::St
     RCLCPP_ERROR(
       node_->get_logger(), "Failed to load image file %s for reason: %s",
       loadParameters.image_file_name.c_str(), e.what());
+    response->result = nav2_msgs::srv::LoadMap::Response::RESULT_INVALID_MAP_DATA;
+    return false;
+  }
+  return true;
+}
+
+nav2_util::CallbackReturn OccGridLoader::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(node_->get_logger(), "OccGridLoader: Configuring");
+
+  // initialize Occupancy Grid msg - needed by loadMapFromYaml
+  msg_ = std::make_unique<nav_msgs::msg::OccupancyGrid>();
 
-    throw std::runtime_error("Failed to load map image file.");
+  if (!loadMapFromYaml(yaml_filename_)) {
+    throw std::runtime_error("Failed to load map yaml file: " + yaml_filename_);
   }
 
-  // Create a service callback handle
+  // Create GetMap service callback handle
   auto handle_occ_callback = [this](
     const std::shared_ptr<rmw_request_id_t>/*request_header*/,
     const std::shared_ptr<nav_msgs::srv::GetMap::Request>/*request*/,
     std::shared_ptr<nav_msgs::srv::GetMap::Response> response) -> void {
-      RCLCPP_INFO(node_->get_logger(), "OccGridLoader: Handling map request");
+      RCLCPP_INFO(node_->get_logger(), "OccGridLoader: Handling GetMap request");
       response->map = *msg_;
     };
 
   // Create a service that provides the occupancy grid
   occ_service_ = node_->create_service<nav_msgs::srv::GetMap>(service_name_, handle_occ_callback);
 
+  // Create the load_map service callback handle
+  auto load_map_callback = [this](
+    const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+    const std::shared_ptr<nav2_msgs::srv::LoadMap::Request> request,
+    std::shared_ptr<nav2_msgs::srv::LoadMap::Response> response) -> void {
+      RCLCPP_INFO(node_->get_logger(), "OccGridLoader: Handling LoadMap request");
+      // Check if map_id is a file
+      if (request->type != nav2_msgs::srv::LoadMap::Request::TYPE_FILE) {
+        RCLCPP_ERROR(node_->get_logger(),
+          "OccGridLoader: unsupported FILE_TYPE in request, can't load map");
+        response->result = nav2_msgs::srv::LoadMap::Response::RESULT_INVALID_TYPE;
+        return;
+      }
+      // Load from file
+      if (loadMapFromYaml(request->map_id, response)) {
+        response->map = *msg_;
+        response->result = nav2_msgs::srv::LoadMap::Response::RESULT_SUCCESS;
+        // if in ACTIVE state, publish map
+        if (node_->get_current_state().id() == lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE) {
```

Comment:
This node state check should happen first thing in the method, right now it doesn't cover the case of downloading the map (whenever that's added). No point in loadMapFromYaml if not active yet

Reply:
Agree. I fixed this in https://github.com/ros-planning/navigation2/pull/1303/commits/aad70fffc1ca75b888647b077986dbae3a09a7b4 and also added a check in the "GetMap" service handler too.